nautilus-typeahead: init at 46.2

bryango opened the pull request:
## Description of changes

- AUR package: https://aur.archlinux.org/packages/nautilus-typeahead
- Also, this revives: #8704 but for gtk4

This package provides a popular variant of the Nautilus file browser that restores the removed "typeahead"-find feature.

In newer versions of Nautilus, typing in the files view will trigger a recursive search under the current directory, which may be rather slow and unintuitive. `nautilus-typeahead` restores the traditional behavior, where typing simply focuses on a matching file under the current view.

This package is an unofficial fork of the official GNOME Nautilus app, published in the Arch User Repository (AUR). To maximize coherence and auditability, we first generate a human-readable patch against the official source, and then apply the patch by overriding the official nautilus package in nixpkgs. More specially,

- `patch.nix` is a fixed output derivation to generate a small, human-readable patch between the nixpkgs nautilus source and the AUR fork.
- `package.nix` applies the patch to the official source by overriding the `nautilus` derivation.
- `update.sh` is used as `passthru.updateScript` to automate the update process.

As an unofficial package, `nautilus-typeahead` is obviously prone to breakages due to upstream API changes. I have hence set myself as the sole maintainer of this package. In particular, gnome and nautilus maintainers need not be responsible for possible (likely) future breakages of this package. However, due to its popularity, I do believe it would be nice to have it in nixpkgs.

<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [x] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

bryango committed:
nautilus-typeahead: init at 46.2

AUR package: https://aur.archlinux.org/packages/nautilus-typeahead

This package provides a popular variant of the Nautilus file browser
that restores the removed "typeahead"-find feature.

In newer versions of Nautilus, typing in the files view will trigger a
recursive search under the current directory, which may be rather slow
and unintuitive. `nautilus-typeahead` restores the traditional behavior,
where typing simply focuses on a matching file under the current view.

This package is an unofficial fork of the official GNOME Nautilus app,
published in the Arch User Repository (AUR). To maximize coherence and
auditability, we first generate a human-readable patch against the
official source, and then apply the patch by overriding the official
nautilus package in nixpkgs. More specially,

- `patch.nix` is a derivation to generate a small, human-readable patch
  between the nixpkgs nautilus source and the AUR fork.
- `package.nix` applies the patch to the official source by overriding
  the `nautilus` derivation.
- `update.sh` is used as `passthru.updateScript` to automate the update
  process.

As an unofficial package, `nautilus-typeahead` is obviously prone to
breakages due to upstream API changes. I have hence set myself as the
sole maintainer of this package. In particular, gnome and nautilus
maintainers need not be responsible for possible (likely) future
breakages of this package. However, due to its popularity, I do believe
it would be nice to have it in nixpkgs.

JohnRTitor commented:
Result of `nixpkgs-review pr 335779` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>3 packages failed to build:</summary>
  <ul>
    <li>nautilus-typeahead</li>
    <li>nautilus-typeahead.dev</li>
    <li>nautilus-typeahead.devdoc</li>
  </ul>
</details>

JohnRTitor commented on pkgs/by-name/na/nautilus-typeahead/patch.nix:
```diff
@@ -0,0 +1,45 @@
+    set -x
+
+    git clone --filter=blob:none --single-branch \
+      https://gitlab.gnome.org/albertvaka/nautilus.git
+
+    cd nautilus
+    git fetch origin "${finalAttrs.rev}"
+    git remote add upstream \
+      https://gitlab.gnome.org/GNOME/nautilus.git
+
+    git fetch --filter=blob:none --tags upstream
+    git format-patch "${nautilus.version}".."${finalAttrs.rev}" --output="$out"
```
This is a very dirty way of doing this. And I am not sure if we wanna even want this fork in Nixpkgs. CC @NixOS/gnome

jtojnar replied:
I do not want to maintain this and each Nautilus bump will break it.

bryango replied:
> This is a very dirty way of doing this.

I understand that this is far from ideal. The reason is, unlike github, AFAIK gitlab cannot generate a `.patch` or `.diff` file which we can `fetchpatch`. That is the reason for doing this manually in `patch.nix`.

> I do not want to maintain this and each Nautilus bump will break it.

Yes I understand. I have stated in the PR,

> As an unofficial package, `nautilus-typeahead` is obviously prone to breakages due to upstream API changes. I have hence set myself as the sole maintainer of this package. In particular, gnome and nautilus maintainers need not be responsible for possible (likely) future breakages of this package. However, due to its popularity, I do believe it would be nice to have it in nixpkgs.

Moreover, I am fine for this PR to be closed if it proves to be a maintenance burden. Simply leaving this in the nixpkgs closed PR list may help others who want the functionality like me.

jtojnar replied:
Main issue I see is that the FOD is coupled to `nautilus.version` attribute. So you might not even get a build failure, you might get a FOD mismatch or a possibly incorrect patch. This kind of spooky action at a distance is confusing and cursed.

You could mitigate it by adding `assert nautilus.version == removeRev finalAttrs.version;` to `outputHash` or including `nautilus.version` in the name.

bryango replied:
> Main issue I see is that the FOD is coupled to `nautilus.version` attribute. So you might not even get a build failure, you might get a FOD mismatch or a possibly incorrect patch. This kind of spooky action at a distance is confusing and cursed.
> 
> You could mitigate it by adding `assert nautilus.version == removeRev finalAttrs.version;` to `outputHash` or including `nautilus.version` in the name.

Indeed, this could be rather dangerous. I have hence decoupled the patch and the nixpkgs nautilus. Instead of `assert`ing I have added a `lib.warnIf` when there is a version mismatch. See the latest commit.

bryango commented:
> Result of `nixpkgs-review pr 335779` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
> 3 packages failed to build

The failure message is evident: there is a hash mismatch from the FOD, that it because the AUR people have released a new patch. Running the `updateScript` should fix it. I will push a commit soon.

As stated in the review comment, I am actually fine for this PR to be closed if it proves to be a maintenance burden. Simply leaving this in the nixpkgs closed PR list may help others who want the functionality like me.

jtojnar commented:
> there is a hash mismatch from the FOD, that it because the AUR people have released a new patch. Running the `updateScript` should fix it. I will push a commit soon.

You are pinning the Git commit so it must be something else – either the Nautilus version changed after you made the PR (not likely since it was [last updated](https://github.com/NixOS/nixpkgs/commit/ca3b765cfbb69f5324ce07266ad471522c7423bb) 4 months ago) or the patch generation is non-deterministic.

Try going back to the previous version and compare the patch output with the original hash against the one you get if you set a new hash (set the hash to an empty string and then let Nix tell you the new hash).

bryango commented:
> You are pinning the Git commit so it must be something else – either the Nautilus version changed after you made the PR (not likely since it was [last updated](https://github.com/NixOS/nixpkgs/commit/ca3b765cfbb69f5324ce07266ad471522c7423bb) 4 months ago) or the patch generation is non-deterministic.

You are absolutely right! Sorry for missing this...

> Try going back to the previous version and compare the patch output with the original hash against the one you get if you set a new hash (set the hash to an empty string and then let Nix tell you the new hash).

I did this and indeed, the issue lies in the git version embedded in the patch:

```diff
--- a/rihisznd4n0fmr7gkpylz0h7igjvl4n3-nautilus-typeahead-46.1-1.patch
+++ b/1skpjbnw3zlm2pahhisa8a4nkvz08pb9-nautilus-typeahead-46.1-1.patch
@@ -54,7 +54,7 @@ index f2fbd8dcf..b5e8be9eb 100644
  
      nautilus_window_slot_set_loading (self, FALSE);
 -- 
-2.45.2
+2.46.0
 
 
 From afd5453a72e2a3d1a71a4c0e9e9d27d08961d26d Mon Sep 17 00:00:00 2001
@@ -384,7 +384,7 @@ index 671fd7081..2d815ab95 100644
          </child>
          <child>
 -- 
-2.45.2
+2.46.0
 
 
 From 2a8e9ca50fb03f80751b83ab7ea08d7dfc77787a Mon Sep 17 00:00:00 2001
@@ -446,7 +446,7 @@ index 1050bd8d9..f2700c07a 100644
       * UI, specifically when the toolbar adds or removes it */
      g_object_ref_sink (self->query_editor);
 -- 
-2.45.2
+2.46.0
 
 
 From f5f593bf36c41756a29d5112a10cf7ec70b8eafb Mon Sep 17 00:00:00 2001
@@ -480,5 +480,5 @@ index 0f6b5eff9..eb57a4bf8 100644
      nautilus_application_set_accelerator (app, "win.go-starred", "Favorites");
      nautilus_application_set_accelerator (app, "win.tab-move-left", "<shift><control>Page_Up");
 -- 
-2.45.2
+2.46.0
```
`fetchpatch` would have taken care of this but we cannot use it for local patches... I will try to find a more robust & reproducible version of `git format-patch`, until then I will mark this PR as draft.

bryango committed:
nautilus-typeahead: make patch generation robust

- Generate patch against the upstream `pkgver`,
  instead of the nixpkgs `nautilus.version`

- Specify `--no-signature` for `git format-patch`,
  such that the output is stable against `git` version updates

bryango committed:
nautilus-typeahead: warn if version mismatch

... between the AUR patch and the Nixpkgs Nautilus

bryango commented:
@ofborg eval

bryango committed:
nautilus-typeahead: bump AUR patch

... by running the `updateScript`, namely `./update.sh`.

Changelog: https://aur.archlinux.org/cgit/aur.git/log/?h=nautilus-typeahead
